### PR TITLE
Update Layerwise Recompute pattern matcher for Bart model 

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Generalize the transformer layer start/end pattern for both Bert and Bart models. 

Introduce a matching algorithm between start and end edges, since Encoder-Decoder model might have mismatch between start and end purely based on the topological sort.

Expose layer-wise recompute in the old pytorch API, since DevDiv Bart work is still using this.

### CODE REVIEW

File: orttraining/orttraining/core/optimizer/transformer_layer_recompute.cc [hunk 1]
```diff
@@ -9,42 +9,79 @@
...
-    // Look for start of a transformer layer
-    if ((layernorm_ops.find(node.OpType()) != layernorm_ops.end() ||
-         dropout_ops.find(node.OpType()) != dropout_ops.end()) &&
+    /*
+    Look for start of a transformer layer
+          [LayerNorm|Dropout|Transpose]
+         /       |         |           \       <----- Start of the transformer layer
+       (.)    MatMul    MatMul       MatMul
+    */
+    if ((IsOpType(node, layernorm_ops) || IsOpType(node, dropout_ops) || IsOpType(node, transpose_ops)) &&
```
[the remark this answers is not in the document]
Comment (wezuo) [reply orttraining/orttraining/core/optimizer/transformer_layer_recompute.cc:42]: btw, which Bart model has this tested on?

[the remark this answers is not in the document]
Comment (SherlockNoMad) [reply orttraining/orttraining/core/optimizer/transformer_layer_recompute.cc:42]: I found a common pattern between bert and bart. So these code would work for both models.  

The  Bart model is from latest Sept.4 FairSeq merged repo.

File: orttraining/orttraining/core/optimizer/transformer_layer_recompute.cc [hunk 2]
```diff
@@ -9,42 +9,82 @@
...
   const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
   for (auto node_index : node_topology_list) {
     auto& node = *graph.GetNode(node_index);
 
-    // Look for start of a transformer layer
-    if ((layernorm_ops.find(node.OpType()) != layernorm_ops.end() ||
-         dropout_ops.find(node.OpType()) != dropout_ops.end()) &&
+    /*
+    Look for start of a transformer layer
+          [LayerNorm|Dropout|Transpose]
```
Comment (pengwa) [inline orttraining/orttraining/core/optimizer/transformer_layer_recompute.cc:41]: seems we won't handle the case when we enable megatron, right? then we will have a MegatronF after layernorm, before QKV usually.

File: orttraining/orttraining/core/optimizer/transformer_layer_recompute.cc [hunk 3]
```diff
@@ -167,19 +244,18 @@ void TransformerLayerRecompute::InsertRecomputeNodes(Graph& graph, const std::ve
...
+  std::vector<const NodeArg*> start_edges, end_edges;
+  Status s = IdentifyTransformerLayerEdges(graph, start_edges, end_edges, logger);
 
-  Status s = IdentifyTransformerLayerEdges(graph, start_end_edges, logger);
   if (!s.IsOK()) {
     modified = false;
     return Status::OK();
   }
 
   // by default, apply recompute expect for the last transformer layer
```
Comment (pengwa) [inline orttraining/orttraining/core/optimizer/transformer_layer_recompute.cc:255]: expect  --> except?

File: orttraining/orttraining/core/optimizer/transformer_layer_recompute.cc [hunk 4]
```diff
@@ -9,42 +9,82 @@
...
   GraphViewer graph_viewer(graph);
   const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
   for (auto node_index : node_topology_list) {
     auto& node = *graph.GetNode(node_index);
 
-    // Look for start of a transformer layer
-    if ((layernorm_ops.find(node.OpType()) != layernorm_ops.end() ||
-         dropout_ops.find(node.OpType()) != dropout_ops.end()) &&
+    /*
+    Look for start of a transformer layer
```
Comment (pengwa) [inline orttraining/orttraining/core/optimizer/transformer_layer_recompute.cc:40]: do we have UT coverage?

**Pull request closed without merging**